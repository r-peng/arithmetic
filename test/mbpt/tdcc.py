import numpy as np
norb = 6
na = norb // 2
nb = na

t_leads = 1.
t_hyb = .4
U = 1.
idot = norb // 2

h1 = np.zeros((norb,norb))
for i in range(norb):
    if i < norb - 1:
        dot = (i == idot or i + 1 == idot)
        h1[i, i + 1] = -t_hyb if dot else -t_leads
    if i > 0:
        dot = (i == idot or i - 1 == idot)
        h1[i, i - 1] = -t_hyb if dot else -t_leads
w,v = np.linalg.eigh(h1)
print('E0=',sum(w[:na])*2)
eri = np.zeros((norb,)*4)
eri[idot,idot,idot,idot] = U

class Model:
    def __init__(self):
        self.h1 = h1
        self.eri = eri
        self.ll = idot
        self.lr = norb - self.ll - 1
        self.no = na
    def get_tmatS(self):
        return self.h1
    def get_vmatS(self):
        return np.zeros_like(self.h1)
    def get_umatS(self):
        return self.eri
model = Model()

from pyscf.cc import td_roccd
eris = td_roccd.ERIs_SIAM(model,mo_energy=w,mo_coeff=v)
no, nv = na, norb-na
t = np.zeros((nv,nv,no,no))
l = np.zeros((no,no,nv,nv))
dt = 0.1
nsteps = 10
rdm1,_ = td_roccd.kernel(eris,t,l,dt*nsteps,dt,RK=4)
rdm1 = np.einsum('npq,ip,jq->nij',rdm1,v,v)
tr = np.einsum('njj->n',rdm1)
for i in range(rdm1.shape[0]):
    print(i,2*rdm1[i,idot,idot],2*tr[i])
