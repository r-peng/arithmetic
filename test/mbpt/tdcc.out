E0= -5.560960262062946
time: 0.0000, EE(mH): 0.0, X: 0.35315075839952925
time: 0.1000, EE(mH): 0.0003708500440069429, X: 0.3516392606182504
time: 0.2000, EE(mH): 0.0011786559177551226, X: 0.3486500057561213
time: 0.3000, EE(mH): 0.002406296360391025, X: 0.34425910402998605
time: 0.4000, EE(mH): 0.004026849356186801, X: 0.33859959572807097
time: 0.5000, EE(mH): 0.006002151396344857, X: 0.33187872918094896
time: 0.6000, EE(mH): 0.008281469923154816, X: 0.3243919422983878
time: 0.7000, EE(mH): 0.010800832052737519, X: 0.31652596291030294
time: 0.8000, EE(mH): 0.013483532847580193, X: 0.3087427617811175
time: 0.9000, EE(mH): 0.016242224633700175, X: 0.3015378870925744
time: 1.0000, EE(mH): 0.018982766889052982, X: 0.29537272298465633
0 (0.9978620127894855-4.83539345851958e-18j) (6.000000000000002-5.46168829623043e-18j)
1 (0.9915154443371537+2.6416839809947543e-18j) (6.000000000000003+8.960541518960306e-19j)
2 (0.9811599162329717+6.320560038448349e-18j) (6.0000000000000036+1.8964756669113468e-17j)
3 (0.9671160306224286+9.396984260986246e-18j) (6.0000000000000036+1.3674949886726654e-17j)
4 (0.9498082351488369-1.3765973505600578e-17j) (6.000000000000004-2.1227280932903207e-17j)
5 (0.9297427595700752-1.3579632210380943e-17j) (6.000000000000003-2.040587073230335e-17j)
6 (0.907482302821623+2.1094520424773717e-17j) (6.0000000000000036+1.390531637860022e-17j)
7 (0.8836193135698323-6.193540629381306e-18j) (6.0000000000000036+8.265771015781087e-18j)
8 (0.8587496931446635-5.664908725826278e-18j) (5.999999999999999+2.6818757175965657e-17j)
9 (0.8334485705603957+3.6591823321385775e-17j) (6.0000000000000036+4.472333961502706e-17j)
10 (0.8082494867985945+5.502516527005867e-18j) (6.000000000000002+3.7133924407628535e-18j)
